Firestore/Source/Core/FSTFirestoreClient.mm

```diff
@@ -157,6 +157,11 @@ - (instancetype)initWithDatabaseInfo:(const DatabaseInfo &)databaseInfo
   return self;
 }
 
+- (void)dealloc {
+  // Clear all delayed operations so that GCD doesn't try to run them after the queue is destroyed.
+  [self.workerDispatchQueue clear];
```

Comment:
How does this interact with tasks scheduled into the future? The scenario I'm thinking of is:
  1. User has active listens
  2. Cancels them all
  3. Deletes the Firestore instance.
The problem is that at (2) we kick off an idle timer on the watch stream. There's a parallel problem with the write stream.

Reply:
If you don't mind, I'd like to discuss this out of band, because I think it's the most significant one.

The idea is to cancel any delayed operations that might still be on the queue. This is done specifically for the libdispatch implementation. In C++, the queue owns both the worker thread and the operations, so the destructor will just clear the operations container and join the thread.  With libdispatch, it's impossible to cancel any operation once it's dispatched (or clear the queue, or even suspend it -- apparently libdispatch crashes if a queue is being destroyed while it's in suspended state). IIUC, it can't be guaranteed that the dispatch queue gets released in a predictable manner, so my concern is that libdispatch may get to run a delayed operation after Firestore instance is destroyed. I have this local test (don't know if it's possible to do this in a more reasonable manner, without relying on `sleep`):
```cpp
  bool called = false;
  {
    AsyncQueue other_queue{std::unique_ptr<internal::Executor>(GetParam()())};
    other_queue.Enqueue([&] {
      other_queue.EnqueueAfterDelay(AsyncQueue::Milliseconds(300), kTimerId1,
                                    [&called] { called = true; });
    });
    other_queue.Clear(); // Comment or uncomment
  }
  std::this_thread::sleep_for(std::chrono::seconds(1));
  EXPECT_FALSE(called);
```
With a call to `Clear`, everything works fine. Without a call to `Clear`, there's a data race when libdispatch gets to run the operation after `other_queue` is destroyed. The way I wrote `AsyncQueue` checks is that the lambda passed to `EnqueueAfterDelay` is wrapped into another lambda that contains a reference to the `AsyncQueue` (this is to avoid `Executor` being aware of the `AsyncQueue` logic). Even if that were not the case, the `DelayedOperation` will try to remove itself from the `Executor`, which is also destroyed at this point. Even if that was somehow avoided, I suspect that any objects that the operation references are probably gone by this point.

What are your thoughts on this? Can Firestore instance be destroyed and recreated during the lifetime of the app? If this can only happen on shutdown, perhaps we don't care? (but it seems it may sometimes lead to a crash on access to deallocated memory, so crashing on shutdown still seems like a problem).

For the scenario you describe, `clear` will cancel the idle timer. Sorry, I presumed  streams don't outlive the Firestore instance, sounds like it's not the case? Can you explain in more detail how they interact? Can Firestore instance be recreated?

Reply:
To summarize our offline discussion:
* the summary is that `clear` should not be needed;
* classes that schedule delayed operations are already supposed to clear them on shutdown, so clearing them one more time is redundant and may mask logic errors. It might be worthwhile to add an assertion, though.
* shutdown process is dispatched on the queue, so the in-progress operation during the shutdown is guaranteed to be the shutdown itself;
* this only leaves any other immediate operations that might be enqueued right after shutdown completes. This behavior is not currently supported and would be a user error; it would lead to app crashing. Supporting that case is not bad, but not high-priority and can be postponed. Implementing it would be a little cumbersome because it's undesirable to wait for potentially long-running operations to finish. The way to go would be to turn those into no-ops, similar to delayed operations, which means they would have to be stored, etc.
* in short, the existing code base either should not get into problematic situations, or else not calling `clear` will not make the situation worse than what is currently in place.